h2o-algos/src/main/java/hex/glrm/GLRM.java

```diff
@@ -1316,6 +1427,82 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
     }
   }
 
+  // Perform de-standardization over matrix A for non NA values only
+  private static class DeStandardizeA extends MRTask<DeStandardizeA> {
+    // Input
+    final int _ncolA;         // Number of cols in training frame
+    final int _ncolX;         // Number of cols in X (k)
+    final int _ncats;         // Number of categorical cols in training frame
+    final double[] _normSub;  // For standardizing training data
+    final double[] _normMul;
+
+    DeStandardizeA(int ncolA, int ncolX, int ncats, double[] normSub, double[] normMul) {
+      assert ncats <= ncolA;
+      _ncolA = ncolA;
+      _ncolX = ncolX;
+      _ncats = ncats;
+
+      _normSub = normSub;
+      _normMul = normMul;
+    }
+
+    @SuppressWarnings("ConstantConditions")  // The method is too complex
+    @Override public void map(Chunk[] cs) {
+      assert _ncolA <= cs.length;
+      if ((_normSub == null) || (_normMul == null)) // do nothing if subtraction or multiplication vectors are null
+        return;
+
+      for (int row = 0; row < cs[0]._len; row++) {
+        // Standardization is performed on numeric columns
+        for (int j = _ncats; j < _ncolA; j++) {
+          double a = cs[j].atd(row);
+          if (Double.isNaN(a)) continue;   // Skip missing observations in row
+
+          int js = j-_ncats;              // perform de-standization for numerical values
+          cs[j].set(row, (a * _normMul[js] + _normSub[js]));
+        }
+      }
+    }
+  }
+
+  // Perform standardization over matrix A for non NA values only
+  private static class StandardizeA extends MRTask<StandardizeA> {
+    // Input
+    final int _ncolA;         // Number of cols in training frame
+    final int _ncolX;         // Number of cols in X (k)
+    final int _ncats;         // Number of categorical cols in training frame
+    final double[] _normSub;  // For standardizing training data
+    final double[] _normMul;
+
+    StandardizeA(int ncolA, int ncolX, int ncats, double[] normSub, double[] normMul) {
+      assert ncats <= ncolA;
+      _ncolA = ncolA;
+      _ncolX = ncolX;
+      _ncats = ncats;
+
+      _normSub = normSub;
+      _normMul = normMul;
+    }
+
+    @SuppressWarnings("ConstantConditions")  // The method is too complex
+    @Override public void map(Chunk[] cs) {
```

Comment:
It appears that this method modifies the original data frame, which is problematic.
For instance, the user might want to do something with the frame while the model is being trained; or maybe the model terminates unexpectedly -- and then we never get to "denormalization" step -- in this case the use will see that running GLRM has "broken" their data.... Finally, there might be small round-off errors introduced during normalization/denormalization, which may matter for those pesky humans.

I would prefer that this method created modified copies of the original Vecs that need to be modified, and left the original Vecs intact.